### PR TITLE
Multithreaded Correctness and Callback support

### PR DESCRIPTION
**Pull request by JRProd**
Fixes issues found by Find-Bugs associated with multithreaded correctness relating to SecretKeys.

Also adds support for adding a callback listener to `commit()` and `apply()` inside the `SharedPreferenceVault` to improve their functionality. `commit()` and `apply()` could fail during encryption and results would become useless. This attempts to fix this issue by giving the user more control over the failure of these two methods.

### CODE REVIEW

File: AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+                    public void onComplete(boolean success) {
+                        assertTrue("Error when applying to sharedPrefs", success);
+                    }
+                })
+        ).edit();
+
+        editor.putString(TEST_STRING_KEY, TEST_STRING_VALUE).apply();
+    }
+
+    public void testNullTypedBundleFailureCommit() {
```
Comment (jaredstockton) [inline AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java]: @lazysp333 This test case should be removed since you can't simulate a null typed bundle

File: AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                    public void onComplete(boolean success) {
+                        assertFalse("Data was added to the SharedPref", success);
+                    }
+                })
+        ).edit();
+
+        editor.commit();
+    }
+
+    public void testNullTypedBundleFailureApply() {
```
Comment (jaredstockton) [inline AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java]: @lazysp333 This test case should be removed since you can't simulate a null typed bundle

File: AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+        SharedPreferences.Editor editor = mSharedPreferenceVault.setSharedPrefVaultWriteListener(
+                new SharedPrefVaultWriteListenerExtension(new OnComplete() {
+                    @Override
+                    public void onComplete(boolean success) {
+                        assertFalse("The key is still initialized", success);
+                    }
+                })
+        ).edit();
+
+        editor.putString(TEST_STRING_KEY, TEST_STRING_VALUE).commit();
```
Comment (jaredstockton) [inline AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java:124]: @lazysp333 Should this be `editor.putString(null, TEST_STRING_VALUE).commit();`

File: AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        SharedPreferences.Editor editor = mSharedPreferenceVault.setSharedPrefVaultWriteListener(
+                new SharedPrefVaultWriteListenerExtension(new OnComplete() {
+                    @Override
+                    public void onComplete(boolean success) {
+                        assertFalse("The key is still initialized", success);
+                    }
+                })
+        ).edit();
+
+        editor.putString(TEST_STRING_KEY, TEST_STRING_VALUE).apply();
```
Comment (jaredstockton) [inline AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java:139]: @lazysp333 Should this be `editor.putString(null, TEST_STRING_VALUE).apply();`

File: README.md [hunk 5]
```diff
@@ -158,9 +159,47 @@ You can use the SharedPreferenceVault with SecretKey generated entirely from the
         SharedPreferenceVaultRegistry.getInstance().getVault(VAULT_ID).rekeyStorage(secretKey);
 
 #### Rekeying
-The vault can be rekeyed at any time. This will delete all values in the shared 
-preference file. This is completely irreversible.
+The vault can be rekeyed at any time. This will delete all values in the shared preference file. This is completely irreversible.
+
+#### API Changes to Commit and Apply
+Typical results of the SharedPreferenc commit and apply are slightly modified in Vault resulting of the extra encryption that is taking place. Because of this, commit and apply can both fail for reasons other than the SharedPreference write failing. The return type for commit is now more important. The recommendation is to **enable exceptions** when creating a SharedPreferenceVault.
```
Comment (jaredstockton) [inline README.md]: @lazysp333 Suggested changes:
Typical behavior of the SharedPreference commit and apply methods are slightly modified in Vault as a result of the added encryption. Because of this, commit and apply can both fail for reasons other than the SharedPreference write failing. This means that you should pay attention to the return value for commit and react appropriately in the case where it returns false. It is recommended to **enable exceptions** when creating a SharedPreferenceVault if you want to see more information on failures (by catching the thrown exception yourself)

File: README.md [hunk 6]
```diff
@@ -158,9 +159,47 @@ You can use the SharedPreferenceVault with SecretKey generated entirely from the
...
+
+These all have the ability to throw `GeneralSecurityException` or `UnsupportedEncodingException` with **exceptions enabled**. These exceptions can occur before the commit or apply is actually attempted and can result in a failure to write data to the SharedPreference file.
+
+1. If using `commit()` consider **enabling exceptions** to allow handling of these errors. Also, pay attention to the return type from `commit()` to make sure data was successfully written.
+2. If using `apply()`, there is a new `SharedPrefVaultWriteListener` interface that can be used to handle callback should any of the errors mentioned above. The `SharedPreferenceVault` interface now supports a new method to add this Listener to the Vault. `setSharedPrefVaultWriteListener(SharedPrefVaultWriteListener listener)` This method can be chained but must be called **before** an editor is created.
+    * The actual result of the write is still ignored when using the new `apply()` with a listener. The listener is only to show potential encryption issues that happen *before* the write.
+    * Setting no listener will behave like the standard `SharedPreference.apply()`. All error are ignored.
```
Comment (jaredstockton) [inline README.md]: @lazysp333 nit: Typo. Change last sentence to All errors are ignored.

File: README.md [hunk 7]
```diff
@@ -158,9 +159,47 @@ You can use the SharedPreferenceVault with SecretKey generated entirely from the
...
+    * The listener will only provide basic information on success and fail, it is still recommended that if more information is required, to **enable exceptions**.
```
Comment (jaredstockton) [inline README.md]: @lazysp333 nit: Suggested changes:
* The listener will only provide basic information on success and fail. It is recommended to **enable exceptions** if more information on the failure is required.

File: README.md [hunk 8]
```diff
@@ -171,6 +210,6 @@ Automated testing tools are often built to trigger on potential cryptographic mi
 This project must be built with gradle. 
 
 *   Version Numbering - The version name should end with "-SNAPSHOT" for non release builds. This will cause the resulting binary, source, and javadoc files to be uploaded to the snapshot repository in Maven as a snapshot build. Removing snapshot from the version name will publish the build on jcenter. If that version is already published, it will not overwrite it.
-*   Execution - To build this libarary, associated tasks are dynamically generated by Android build tools in conjunction with Gradle. Example command for the production flavor of the release build type: 
+*   Execution - To build this library, associated tasks are dynamically generated by Android build tools in conjunction with Gradle. Example command for the production flavor of the release build type:
```
Comment (jaredstockton) [inline README.md:213]: @lazysp333 Nice catch!

**Review by jaredstockton — CHANGES_REQUESTED** [review]
Nice job overall! After addressing the review comments, my only other comment is to break up the multithreaded correctness and callback support into two separate PRs, since they are independent of each other. My apologies for not mentioning that sooner but it should be relatively trivial :)

File: AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java (hunk 3 above)
Comment (JRProd) [reply AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java:124]: @dev777 This is the null secret key, AKA the user is trying to encrypt data without having a secret key set. That is why I clear the storage before.

If you would also like, I can add a test case for a attempt to set a null shared pref key as well

Comment (jaredstockton) [reply AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java:124]: @lazysp333 I think that renaming this test case to `testNullSecretKeyFailureCommit` and then adding the test case for null shared pref key makes sense. Same recommendation for the [testNullKeyFailureApply comment](https://github.com/BottleRocketStudios/Android-Vault/pull/14/files/6488f1fa681e3e4a8052da8f1d0ab886a5d73241#r205763149) as well

Comment (JRProd) [reply AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java:124]: @dev777 Another thing, putting a null value in this IS acceptable in its current state, the `editor.commit()` returns true (same with apply). Is this expected behavior?

Comment (JRProd) [reply AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java:124]: These have been explained and fixed

File: AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java (hunk 4 above)
Comment (JRProd) [reply AndroidVault/vault/src/androidTest/java/com/bottlerocketstudios/vault/test/TestSharedPrefListener.java:139]: These have been explained and fixed
